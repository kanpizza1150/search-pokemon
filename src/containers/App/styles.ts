import styled from 'styled-components'
import bg from '../../asset/images/bg.webp'
import {
  setFlex,
  setColor,
  borderStyle,
  setTransition,
} from '../../utils/globalStyles'

export const Container = styled.div`
  background: url(${bg}) no-repeat fixed center;
  min-height: 100vh;
  padding: 2rem;
  display: grid;
  grid-template-rows: 10rem 1fr;
  .search {
    &-wrapper {
      ${setFlex()};
    }
  }
  .body {
    ${setFlex()};
  }

  .search {
    &__wrapper {
      background-color: ${setColor.white};
      ${borderStyle};
      max-width: 50rem;
      width: 100%;
      ${setFlex('space-between')};
      input {
        padding: 1.5rem;
        font-size: 2rem;
        background-color: ${setColor.transparent};
        border: none;
        outline: none;
        width: 100%;
      }
    }
  }
  .search__button {
    border: none;
    outline: none;
    width: 2rem;
    font-size: 2rem;
    margin-right: 1.5rem;
    background: ${setColor.transparent};
    cursor: pointer;
    color: ${setColor.grey};
    ${setTransition()};
    &:hover {
      color: ${setColor.black};
    }
  }
`
export const LoadingWrapper = styled.div`
  ${setFlex()};
  flex-direction: column;
`
