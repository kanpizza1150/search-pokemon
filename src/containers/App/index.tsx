import { useState, useEffect } from 'react'
import { useQuery } from '@apollo/client'
import { POKEMONS_ALL } from '../../utils/queries'
import { Container } from './styles'
import Table from '../../components/Table'
import { Pokemon } from '../../components/Card'
const App: React.FC = () => {
  const { error, data } = useQuery(POKEMONS_ALL)
  const [searchKey, setSearchKey] = useState<string>('')
  const [pokemons, setPokemons] = useState<[]>([])

  useEffect(() => {
    if (data) {
      setPokemons(data.pokemons)
    }
  }, [data])

  const filterPokemon = (key): void => {
    const result = pokemons.filter((pokemon: Pokemon) =>
      pokemon.name.toLowerCase().startsWith(key)
    )
    console.log(`result`, result)
  }

  const onSearch = (e: any): void => {
    const { value } = e.target
    setSearchKey(value)
    filterPokemon(value)
  }

  const _renderLoadingState = () => <>Loading...</>
  const _renderErrorState = () => <>Fetching data error</>

  const _handleState = (): JSX.Element => {
    let items: JSX.Element = _renderLoadingState()
    if (error) {
      items = _renderErrorState()
    } else if (data) {
      items = <Table pokemons={pokemons} />
    }
    return items
  }

  return (
    <Container>
      Pokemon
      <input
        onChange={onSearch}
        value={searchKey}
        placeholder='Search by name'
      />
      {_handleState()}
    </Container>
  )
}
export default App
