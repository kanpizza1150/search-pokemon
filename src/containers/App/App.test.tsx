import { fireEvent, render } from '@testing-library/react'
import { useRef } from 'react'
import App from './index'
const mokedPokemons = [{ name: 'a' }, { name: 'b' }]
let mockedUseQuery = {
  loading: false,
  data: {},
  error: {},
}
let mockedUseLocation = {}

let mockedUseHistory = {
  push: jest.fn(),
}

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useRef: jest.fn(),
}))

jest.mock('@apollo/client', () => ({
  useQuery: () => mockedUseQuery,
  gql: jest.fn(),
  ApolloProvider: jest.fn((props) => (
    <apolloprovider>{props.children}</apolloprovider>
  )),
}))
jest.mock('react-router-dom', () => ({
  useLocation: () => mockedUseLocation,
  useHistory: () => mockedUseHistory,
}))
jest.mock('../../components/List', () => ({ children }) => (
  <loader data-testid='loader'>{children}</loader>
))
jest.mock('../../components/List', () => ({ children }) => (
  <list data-testid='list'>{children}</list>
))

describe('App Container', () => {
  beforeEach(() => {
    jest.restoreAllMocks()
    jest.resetAllMocks()
    mockedUseQuery.loading = false
    mockedUseQuery.data = {}
    mockedUseQuery.error = false
    useRef.mockReturnValue({ current: 0 })
  })
  it('Should match snapshot', () => {
    const { asFragment } = render(<App />)
    expect(asFragment).toMatchSnapshot()
  })
  it('Should render loading state', () => {
    mockedUseQuery.loading = true
    const { getByTestId } = render(<App />)
    expect(getByTestId('loader')).toBeInTheDocument()
  })
  it('Should render error state', () => {
    mockedUseQuery.error = true
    const { getByTestId } = render(<App />)
    expect(getByTestId('error-message').textContent).toBe('Fetching data error')
  })
  describe('Search section', () => {
    it('Should render search input', () => {
      const { getByTestId } = render(<App />)
      expect(getByTestId('search-input').placeholder).toBe(
        "SEARCH BY POKEMON'S NAME"
      )
      expect(getByTestId('search-clear').nodeName).toBe('BUTTON')
    })
    it('Should click on clear search', () => {
      const { getByTestId } = render(<App />)
      fireEvent.click(getByTestId('search-clear'))
      expect(mockedUseHistory.push).toBeCalledWith({ search: `` })
    })
    it('Should type on Search input', () => {
      mockedUseQuery.data.pokemons = mokedPokemons
      const { getByTestId } = render(<App />)
      fireEvent.change(getByTestId('search-input'), {
        target: { value: 'test' },
      })
      expect(mockedUseHistory.push).toBeCalledWith({ search: `?name=test` })
    })
  })
  describe('Scroll', () => {
    it('Should type on Search input', () => {
      mockedUseQuery.data.pokemons = mokedPokemons
      render(<App />)
      fireEvent.scroll(window, { target: { scrollY: 2000 } })
      expect(useRef).toBeCalledWith(10)
      fireEvent.scroll(window, { target: { scrollY: 3000 } })
      expect(useRef).toBeCalledWith(20)
    })
  })
})
