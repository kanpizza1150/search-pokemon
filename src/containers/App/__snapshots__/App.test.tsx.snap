// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Container Should match snapshot 1`] = `[Function]`;
